### PR TITLE
Reinstate: SwiftDriver: initial work to properly handle android cross-compilation

### PR DESCRIPTION
**Pull request by hyp**
This PR reinstates https://github.com/swiftlang/swift-driver/pull/1560, after it was reverted in https://github.com/swiftlang/swift-driver/pull/1662. 

The revert happened because of a unit test failure in the `SwiftDriverTests.SwiftDriverTests testRelativeResourceDir` test. This PR fixes this test failure by ensuring that SDKROOT is reset before the test is run, and creates a sibling test that reflects the new behavior.

### CODE REVIEW

**Comment by hyp** [comment]
@swift-ci please test

File: Sources/SwiftDriver/Jobs/GenericUnixToolchain+LinkerSupport.swift [hunk 1]
```diff
@@ -147,14 +147,24 @@ extension GenericUnixToolchain {
...
-        let swiftrtPath = VirtualPath.lookup(targetInfo.runtimeResourcePath.path)
-          .appending(
-            components: targetTriple.platformName() ?? "",
-            String(majorArchitectureName(for: targetTriple)),
-            "swiftrt.o"
-          )
-        commandLine.appendPath(swiftrtPath)
+        let rsrc: VirtualPath
+        // Prefer the swiftrt.o runtime file from the SDK if it's specified.
+        if let sdk = targetInfo.sdkPath.flatMap({ VirtualPath.lookup($0.path) }) {
```
Comment (compnerd) [inline Sources/SwiftDriver/Jobs/GenericUnixToolchain+LinkerSupport.swift:160]: Nit:

```suggestion
        if let sdk = targetInfo.sdkPath, let sdk = VirtualPath.lookup(sdk.path) {
```

might be a bit more accessible?

**Review by compnerd — APPROVED** [review]

**Comment by hyp** [comment]
@swift-ci please test

**Review by artemcm — APPROVED** [review]

**Comment by compnerd** [comment]
@swift-ci please test Windows platform

**Pull request merged**